src/fdcache.cpp

```diff
@@ -338,13 +319,64 @@ bool FdManager::HaveLseekHole()
         }
     }
     close(fd);
-    unlink(TMPFILE_FOR_CHECK_HOLE);
 
     FdManager::checked_lseek   = true;
     FdManager::have_lseek_hole = result;
     return FdManager::have_lseek_hole;
 }
 
+bool FdManager::SetTmpDir(const char *dir)
+{
+    if(!dir || '\0' == dir[0]){
+        tmp_dir = "/tmp";
+    }else{
+        tmp_dir = dir;
+    }
+    return true;
+}
+
+bool FdManager::IsDir(const std::string* dir)
+{
+    // check the directory
+    struct stat st;
+    if(0 != stat(dir->c_str(), &st)){
+        S3FS_PRN_ERR("could not stat() directory %s by errno(%d).", dir->c_str(), errno);
+        return false;
+    }
+    if(!S_ISDIR(st.st_mode)){
+        S3FS_PRN_ERR("the directory %s is not a directory.", dir->c_str());
+        return false;
+    }
+    return true;
+}
+
+bool FdManager::CheckTmpDirExist()
+{
+    if(FdManager::tmp_dir.empty()){
+        return true;
+    }
+    return IsDir(&tmp_dir);
+}
+
+FILE* FdManager::MakeTempFile() {
+    int fd;
+    char cfn[PATH_MAX];
+    std::string fn = tmp_dir + "/s3fstmp.XXXXXX";
+    strncpy(cfn, fn.c_str(), sizeof(cfn) - 1);
+    cfn[sizeof(cfn) - 1] = '\0';
+
+    fd = mkstemp(cfn);
```

Comment:
It would be fine to just call `mkstemp(fn.c_str())` here.

Reply:
`mkstemp()` needs a writable buffer to return the filename (the changed template). 
I'm not quite sure, but I think `fn.c_str()` can't be used here because it's not supposed to be changed.

Reply:
Sorry, you are correct -- thanks for clarifying!